### PR TITLE
Unpack dictionary parameters

### PR DESCRIPTION
**Pull request by bpmeek**
## Description
In response to this [ticket](https://github.com/kedro-org/kedro/issues/2749)

## Development notes
Can now unpack dictionaries in the `inputs` argument of a node, example below.

_unpack_params added to `kedro/kedro/pipeline/node.py`, this function iterates over node inputs and updates `node.inputs` if needed.

*note*: It is assumed that the dictionary entries and kwarg have the same name

### Changes:

`kedro/kedro/pipeline/modular_pipeline.py` was updated to not throw errors when mapping dataset names.
`_is_single_parameter` returns true if name starts with `**params:`
`_normalize_param_name` does not append `params:` if the name already begins with `**params:`
`_validate_datasets_exist` removes datasets that begin with `**` from `non_existent`, I have confirmed that `kedro/runner/runner.py` will catch these missing datasets.

### Undesirable behavior

Currently modular pipelines will still namespace the unpacked parameters, I'll need assistance with either updating the rules list or adjusting the `rename` function in a way that makes sense.

### Examples:

`nodes.py`
```python
def split_data(data: pd.DataFrame, features: List[str], test_size, random_state) -> Tuple:
    X = data[features]
    y = data["price"]
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=test_size, random_state=random_state
    )
    return X_train, X_test, y_train, y_test
```

`pipeline.py`
```python
node(
    func=split_data,
    inputs=["model_input_table", "**params:model_options"],
    outputs=["X_train", "X_test", "y_train", "y_test"],
    name="split_data_node",
),
```

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [X] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [X] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [X] Updated the documentation to reflect the code changes
- [X] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [X] Added tests to cover my changes
- [X] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Comment by astrojuanlu** [comment]
Thanks for this PR @bpmeek! We'll review it shortly 🙏🏼

File: kedro/pipeline/node.py [hunk 1]
```diff
@@ -681,3 +683,27 @@ def _get_readable_func_name(func: Callable) -> str:
...
+    Returns:
+        Either original inputs if no input was unpacked or a list of inputs if an input was unpacked.
+    """
+    use_new = False
+    new_inputs = []
+    for idx, _input in enumerate(_to_list(inputs)):
+        if _input.startswith("**"):
+            use_new = True
+            dict_root = _input.split(":")[-1]
+            _func_arguments = [arg for arg in inspect.signature(func).parameters]
```
Comment (datajoely) [inline kedro/pipeline/node.py:702]: So there is a soft assumption that the dictionary keys match the function arguments. Do we want to throw a warning or error if:

(a) No arguments match
(b) Partial arguments match

Comment (bpmeek) [reply kedro/pipeline/node.py:702]: So that is an assumption, I suppose we could also implement tuple unpacking if a user wanted to use index instead of key-value pairs.

The current behavior of no match/partial match is consistent I think with what users would expect. If you pass the dictionary `model_options` and the function expects a key `features` and that key is missing you would get the following error before the run.

`ValueError: Pipeline input(s) {'params:model_options.features'} not found in the DataCatalog`

Comment (datajoely) [reply kedro/pipeline/node.py:702]: true! interested to see if others have an opinion though

Comment (ElenaKhaustova) [reply kedro/pipeline/node.py:702]: I would vote for adding at least basic validation because there are definitely cases in which this method will not work correctly, for example, when having a `**kwargs'—only node function. Then, we could also extend tests with negative samples so it's clear which cases we cover and which we do not.

**Review by ElenaKhaustova — COMMENTED** [review]
Thank you for the PR!

Left some comments/suggestion regarding the implementation. My main push is to clarify which cases do we cover by adding basic validation, negative tests and small usage example [here](https://docs.kedro.org/en/stable/nodes_and_pipelines/nodes.html)

File: kedro/pipeline/node.py (hunk 1 above)
Comment (bpmeek) [reply kedro/pipeline/node.py:702]: I'm a little confused about how we would implement any validation at this point in the process, this function is called when the Node object is being initialized and I don't believe the catalog is available to verify against is it?


> there are definitely cases in which this method will not work correctly, for example, when having a `**kwargs`—only node function.

Can you help me understand this?  if your function was something like `def foo(float_1, integer_2, string_3):` you could pass `**params:dictionary` and it would work. If your function was `def foo(**kwargs)` then this implementation wouldn't work but I wasn't under the impression that was what was being asked for, because you could just change it to `def foo(kwargs)` and then access the values normally from there.

Comment (ElenaKhaustova) [reply kedro/pipeline/node.py:702]: Sure. I don't mean that this is the expected usage example and we aim to make it work. But it doesn't mean that someone cannot make that (or something else that breaks the current unpacking method) by mistake if syntax allows. There might be some other cases as well. To simplify the validation, we can follow the approach proposed above - check no/partial argument match and add some possible incorrect test cases to ensure the validation catches them. At the same time extend the docs with the expected usage example.

Comment (astrojuanlu) [reply kedro/pipeline/node.py:702]: I think this is the main outstanding discussion right?

**Comment by noklam** [comment]
@bpmeek Do you want to continue this PR? I am eager to see this being merged finally.

**Comment by bpmeek** [comment]
@noklam, I can continue it, I think I'm just stuck right now on the best way to implement checking for partial matches, since the current process doesn't have access to the catalog when it's populating the required datasets and the verification doesn't know what dictionaries have been unpacked.

File: kedro/pipeline/node.py [hunk 2]
```diff
@@ -681,3 +683,28 @@ def _get_readable_func_name(func: Callable) -> str:
...
+            for param in _func_arguments[idx:]:
+                new_inputs.append(f"params:{dict_root}.{param}")
+else:
+     new_inputs.append(_input)
+        new_inputs.append(_input)
+    if use_new:
+        return new_inputs
+    return inputs
```
Comment (noklam) [inline kedro/pipeline/node.py:710]: This seems to be completely broken when I try to test

Comment (noklam) [reply kedro/pipeline/node.py:710]: Feel free to ping me again if it's ready for review, I cannot fix the error so it's a bit hard to review the logic.

Comment (bpmeek) [reply kedro/pipeline/node.py:710]: It is, I'm pushing a new update, I think it got broken when I accepted a change in the online editor.

File: kedro/pipeline/node.py [hunk 3]
```diff
@@ -681,3 +683,27 @@ def _get_readable_func_name(func: Callable) -> str:
...
+
+def _unpack_params(
+    func: Callable, inputs: None | str | list[str] | dict[str, str]
+) -> None | str | list[str] | dict[str, str]:
+    """Iterate over Node inputs to see if they need to be unpacked.
+
+    Returns:
+        Either original inputs if no input was unpacked or a list of inputs if an input was unpacked.
+    """
+    use_new = False
```
[the remark this answers is not in the document]
Comment (bpmeek) [reply kedro/pipeline/node.py:696]: done!

File: kedro/pipeline/node.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bpmeek) [reply kedro/pipeline/node.py:702]: I think this makes sense and went ahead and made the change, but I'm not sure which would be more efficient, creating a list of every node's function or making the list multiple times _if_ the input is a dictionary.

File: kedro/pipeline/node.py [hunk 4]
```diff
@@ -681,3 +683,27 @@ def _get_readable_func_name(func: Callable) -> str:
...
+    new_inputs = []
+    for idx, _input in enumerate(_to_list(inputs)):
+        if _input.startswith("**"):
+            use_new = True
+            dict_root = _input.split(":")[-1]
+            _func_arguments = [arg for arg in inspect.signature(func).parameters]
+            for param in _func_arguments[idx:]:
```
[the remark this answers is not in the document]
Comment (bpmeek) [reply kedro/pipeline/node.py:703]: I'm not sure I understand what you're recommending.

**Comment by bpmeek** [comment]
Are the additional checks/tests here sufficient or is there still more to be done?

**Comment by astrojuanlu** [comment]
Thanks for your patience @bpmeek! Let us look into this again.

**Comment by noklam** [comment]
I have a long list of PRs so I will only have time to look at this next week earliest.

**Comment by astrojuanlu** [comment]
> Hi @bpmeek , we're slowly catching up with backlog items, sorry about our slow response.
>
> I'm not up to date with https://github.com/kedro-org/kedro/pull/3905 but my understanding is that Kedro-Viz, as consumer of Kedro, shouldn't break - otherwise that would be a sign of a breaking change, that should not have happened until the next Kedro release, 0.20.*.
> 
> In other words, the PR should be backwards compatible.
> 
> The heads up is appreciated 🙏🏼 but for now there's nothing for us to do here.

_Originally posted by @astrojuanlu in https://github.com/kedro-org/kedro-viz/issues/1933#issuecomment-2368645041_

**Review by ElenaKhaustova — CHANGES_REQUESTED** [review]
I'm sorry some comments on the validation required were not clear. I think to move forward with this PR we need to:

1. Address comments regarding the implementation;
2. Update the implementation description and clarify what cases are supported and what are not, so it was clear for the reviewers whether something is skipped intentionally;
3. Add a validation based on 2. and make sure the rest cases are failing gracefully;
4. Extend documentation with the example of added interface: https://docs.kedro.org/en/stable/nodes_and_pipelines/nodes.html#nodes

Thank you @bpmeek for the solution! Please let us know if you want to continue with it or need some help from our side.

File: kedro/pipeline/node.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ElenaKhaustova) [reply kedro/pipeline/node.py:698]: Could we please rename it?

**Comment by astrojuanlu** [comment]
hi @bpmeek do you still intend to pursue this?

**Comment by merelcht** [comment]
This PR has been open for a long time and diverged from the `main`/`develop` branch. I'll close it for now and suggest that if @bpmeek or anyone else wants to continue, they open a new PR. Thanks for being a contributor to Kedro and helping us make the product better. We appreciate your help a lot! ⭐
